### PR TITLE
Bugfix/histogram clears catergory and property selection on graph element selection change #1129

### PR DESCRIPTION
**Pull request by sol695510**
### Description of the Change
Stored the selections for Category, Property and Formatter in a new state within the HistogramState.

### Alternate Designs
N/A

### Why Should This Be In Core?
N/A

### Benefits
Reduces latency of analysis because the user does not need to reselect the options for Category, Property and Formatter every time they switch between graph elements.

### Possible Drawbacks
N/A

### Verification Process
Followed steps given in description of issue.

### Applicable Issues
#1129

### CODE REVIEW

**Comment by arcturus2** [comment]
Thanks @sol695510 it works well. Can you please add a what's new entry for this change to let users know?

**Comment by sol695510** [comment]
> Thanks @sol695510 it works well. Can you please add a what's new entry for this change to let users know?

What's New entry added.

File: CoreWhatsNewView/src/au/gov/asd/tac/constellation/views/whatsnew/whatsnew.txt [hunk 1]
```diff
@@ -1,6 +1,9 @@
 == 3030-12-31 Getting Started
 <p>If you're new to Constellation, read the <a href="" helpId="au.gov.asd.tac.constellation.functionality.gettingstarted">getting started guide</a>.</p>
 
+== 2021-09-10 Histogram View now save options chosen for Category and Property per Graph Element
```
Comment (antares1470) [inline CoreWhatsNewView/src/au/gov/asd/tac/constellation/views/whatsnew/whatsnew.txt:4]: I would simplify the title here to something like "Histogram View"

**Review by antares1470 — CHANGES_REQUESTED** [review]
Found a bug. If you select a Category other than "Attribute" and then try choose a property, it resets the Category back to Attribute. I tested this in master as well and didn't encounter the issue.
histogramStateBug

File: CoreHistogramView/src/au/gov/asd/tac/constellation/views/histogram/HistogramState.java [hunk 2]
```diff
@@ -30,66 +30,45 @@
...
-            this.elementStateMap = Map.of(
-                    GraphElementType.VERTEX, new GraphElementState(),
-                    GraphElementType.TRANSACTION, new GraphElementState(),
-                    GraphElementType.EDGE, new GraphElementState(),
-                    GraphElementType.LINK, new GraphElementState());
             this.elementType = GraphElementType.VERTEX;
-            this.attributeType = AttributeType.ATTRIBUTE;
-            this.attribute = null;
-            this.binFormatter = BinFormatter.DEFAULT_BIN_FORMATTER;
+            this.setElementState();
```
Comment (antares1470) [inline CoreHistogramView/src/au/gov/asd/tac/constellation/views/histogram/HistogramState.java:64]: can omit the `this` keyword here

File: CoreHistogramView/src/au/gov/asd/tac/constellation/views/histogram/HistogramState.java [hunk 3]
```diff
@@ -30,66 +30,45 @@
...
             this.binComparator = BinComparator.REVERSE_KEY;
             this.binSelectionMode = BinSelectionMode.FREE_SELECTION;
         } else {
-            this.elementStateMap = original.elementStateMap;
+            this.elementStateMap = Map.copyOf(original.elementStateMap);
             this.elementType = original.elementType;
-            this.attributeType = original.elementStateMap.get(elementType).getAttributeType();
-            this.attribute = original.elementStateMap.get(elementType).getAttribute();
-            this.binFormatter = original.elementStateMap.get(elementType).getBinFormatter();
+            this.setElementState();
```
Comment (antares1470) [inline CoreHistogramView/src/au/gov/asd/tac/constellation/views/histogram/HistogramState.java:71]: same here

**Review by antares1470 — CHANGES_REQUESTED** [review]
Tested and not seeing that issue anymore

**Review by antares1470 — APPROVED** [review]

**Review by arcturus2 — APPROVED** [review]
Works well. I resolved the merge conflict that came up with `whatsnew.txt` and will merge it in now.

**Pull request merged**